File: easy/p7-reverse.py
'''
给出一个 32 位的有符号整数，你需要将这个整数中每位上的数字进行反转。
'''
'''
给出一个 32 位的有符号整数，你需要将这个整数中每位上的数字进行反转。
'''

class Solution:
    def reverse(self, x):
        sign = 1
        if x < 0:
            x = -x
            sign = -1
        l1 = []
        
        while x > 0:
            l1.append(x%10)
            x //=10
        result = sum([x*(10**(len(l1)-i-1)) for i, x in enumerate(l1)])
        if result >= 2**31:
            result = 0
        return int(result)*sign
    def reverse1(self, x):
        sign = 1
        if x < 0:
            x = -x
            sign = -1
        l1 = []
        while x > 0:
            l1.append(x%10)
            x //=10
        result = 0
        maxBit = 10**(len(l1)-1)
        for i in l1:
            result += i*maxBit
            maxBit /= 10
        if result >= 2**31:
            result = 0
        return int(result)*sign

print(Solution().reverse(123))
# print(Solution().reverse(1534236469))
# print(Solution().reverse(123456789))
# print(Solution().reverse(-12345))

File: easy/p7-reverse/p7-reverse.cpp

#include <cmath>
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int reverseGW(int x) {
        int rev = 0;
        while (x != 0) {
            int pop = x % 10;
            x /= 10;
            if (rev > INT_MAX / 10 || (rev == INT_MAX / 10 && pop > 7)) //0x7F FF FF FF
                return 0;
            if (rev < INT_MIN / 10 || (rev == INT_MIN / 10 && pop < -8)) //0x8F FF FF FF
                return 0;
            rev = rev * 10 + pop;
        }
        return rev;
    }

    int reverse(int ix) {
        int sign = 1;
        long x = ix;
        if (x < 0) {
            sign = -1;
            x *= -1;
        }
        std::vector<int> nums;
        while (true) {
            if (x > 0) {
                nums.push_back(x % 10);
                x /= 10;
            }
            if (x == 0) {
                break;
            }
        }
        unsigned long result = 0;
        for (size_t i = 0; i < nums.size(); i++) {
            result += nums[i] * pow(10, nums.size() - i - 1);
        }

        if (result > 2147483648) {
            result = 0;
        }
        return int(result * sign);
    }
};

int main() {
    Solution test;
    cout << test.reverse(123) << endl;
    cout << test.reverse(-123) << endl;
    cout << test.reverse(2147483647) << endl;  //2^31-1
    cout << test.reverse(-2147483648) << endl; //2^31-1
}
